casatables/src/glue.cc

```diff
@@ -131,8 +133,130 @@ extern "C" {
         return -1;
     }
 
+    // TableDesc
+
+    GlueTableDesc *
+    tabledesc_create(
+        // TODO: figure out exactly what this is.
+        // Apparently this is actually the name of the table?
+        const StringBridge &type
+    )
+    {
+        // TODO: expose this?
+        GlueTableDesc::TDOption tdOption = GlueTableDesc::TDOption::New;
+        return new GlueTableDesc(bridge_string(type), tdOption);
+    }
+
+    GlueTableDesc *
+    tabledesc_add_scalar_column(
+        GlueTableDesc &tableDesc,
+        GlueDataType dataType,
+        const StringBridge &columnName,
+        ExcInfo &exc
+    )
+    {
+        try {
+
+            switch (dataType) {
+
+#define CASE(DTYPE, CPPTYPE) \
+            case casacore::DTYPE: { \
+                tableDesc.addColumn(casacore::ScalarColumnDesc<CPPTYPE>( \
+                    bridge_string(columnName) \
+                )); \
+                break; \
+            }
+
+            CASE(TpBool, casacore::Bool)
+            CASE(TpChar, casacore::Char)
+            CASE(TpUChar, casacore::uChar)
+            CASE(TpShort, casacore::Short)
+            CASE(TpUShort, casacore::uShort)
+            CASE(TpInt, casacore::Int)
+            CASE(TpUInt, casacore::uInt)
+            CASE(TpFloat, float)
+            CASE(TpDouble, double)
+            CASE(TpComplex, casacore::Complex)
+            CASE(TpDComplex, casacore::DComplex)
+
+#undef CASE
+
+            case casacore::TpString:
+                throw std::runtime_error("use table_get_scalar_column_data_string for TpString columns");
+
+            default:
+                throw std::runtime_error("unhandled scalar column data type");
+            }
+        } catch (...) {
+            handle_exception(exc);
+        }
+
+        return &tableDesc;
+    }
+
+    // GlueTableDesc *
+    // tabledesc_add_column_array(
+    //     GlueTableDesc &tableDesc,
+    //     GlueDataType dataType,
+    //     const StringBridge &columnName,
+    //     casacore::Int ndim
+    // )
+    // {
+    //     const casacore::ColumnDesc column = new casacore::ArrayColumnDesc<dataType>(
+    //         bridge_string(columnName),
+    //         ndim
+    //     );
+
+    //     tableDesc.addColumn(column);
+    // }
+
     // Tables
 
+    GlueTable *
+    table_create(
+        const StringBridge &path, 
+        // Description of columns and keys in the table
+        GlueTableDesc &tableDesc,
+        // number of rows
+        unsigned long nrrow,
+        ExcInfo &exc
+    )
+    {
+        // TODO: expose this as an argument?
+        // I think the only options here are New, NewNoReplace and Scratch, but 
+        // not sure whether it's worth extending the TableOpenMode enum or 
+        // doing something else. Open to suggestions
```

Comment:
Assuming that the semantics of the options match well to their names, I think it would be good to at least expose the NewNoReplace mode in one way or another. Analogous to the `O_EXCL` option of the Unix file `open()` call, being able to create a new entity without replacing a pre-existing one is an important capability for avoiding race conditions, especially in parallel-processing environments where you might have several nodes doing processing on a shared filesystem.

Reply:
ok cool, I've implemented `NewNoReplace`. There's another option called `Scratch`, but I have no idea what it's used for so I've left it commented out to show how it might be implemented. Should this be exposed too? Or should I just get rid of the comments? The casacore description says "new table, which gets marked for delete".

Reply:
Usually, a "scratch" mode means that the data are created under a temporary path that's then immediately unlinked on the filesystem. The way that Unix I/O works, you can still write to and read from the file, but when you close it, the OS automatically removes it all. Since MS datasets live in directories, the cleanup may be a bit more involved since I think the above trickiness only works with regular files. Anyway, I personally wouldn't expect that to be a very commonly-used mode. It might be helpful to add a comment somewhere stating that the "scratch" mode exists but is unimplemented in Rubbl. If there's a tidal wave of popular demand, we can add it :-)

Reply:
ok, I've added a link to this discussion in the documentation.